src/libmongoc/src/mongoc/mongoc-interrupt.c

```diff
@@ -0,0 +1,315 @@
+   server_addr_in.sin_addr.s_addr = htonl (INADDR_LOOPBACK);
+   ret = mongoc_socket_bind (listen_socket,
+                             (struct sockaddr *) &server_addr_in,
+                             sizeof (server_addr_in));
+   if (ret == -1) {
+      _log_errno ("bind failed", mongoc_socket_errno (listen_socket));
+      GOTO (fail);
+   }
+
+   ret = mongoc_socket_listen (listen_socket, 1);
+   if (ret == -1) {
+      _log_errno ("listen failed", mongoc_socket_errno (listen_socket));
+      GOTO (fail);
+   }
+
+   sock_len = sizeof (server_addr);
+   ret = mongoc_socket_getsockname (
+      listen_socket, (struct sockaddr *) &server_addr, &sock_len);
+   if (-1 == ret) {
+      _log_errno ("getsockname failed", mongoc_socket_errno (listen_socket));
+      GOTO (fail);
+   }
+
+   interrupt->impl.socket_pair.read =
+      mongoc_socket_new (server_addr.ss_family, SOCK_STREAM, 0);
+   if (!interrupt->impl.socket_pair.read) {
+      MONGOC_ERROR ("socket creation failed");
+      GOTO (fail);
+   }
+
+   /* Begin non-blocking connect. */
+   ret = mongoc_socket_connect (
+      interrupt->impl.socket_pair.read, (struct sockaddr *) &server_addr, sock_len, 0);
+   if (ret == -1 &&
+       !MONGOC_ERRNO_IS_AGAIN (mongoc_socket_errno (interrupt->impl.socket_pair.read))) {
+      _log_errno ("connect failed",
+                  mongoc_socket_errno (interrupt->impl.socket_pair.read));
+      GOTO (fail);
+   }
+
+   interrupt->impl.socket_pair.write = mongoc_socket_accept (
+      listen_socket, bson_get_monotonic_time () + timeout_ms * 1000);
+   if (!interrupt->impl.socket_pair.write) {
+      _log_errno ("accept failed", mongoc_socket_errno (listen_socket));
+      GOTO (fail);
+   }
+
+   /* Create an unowned socket. interrupt_socket has 0 for the pid, so it will
+    * be considered unowned. */
+   interrupt_socket = bson_malloc0 (sizeof (mongoc_socket_t));
+   interrupt_socket->sd = interrupt->impl.socket_pair.read->sd;
+   /* Creating the stream takes ownership of the mongoc_socket_t. */
+   interrupt->stream = mongoc_stream_socket_new (interrupt_socket);
+   success = true;
+fail:
+   mongoc_socket_destroy (listen_socket);
+   if (!success) {
+      _mongoc_interrupt_destroy (interrupt);
+      interrupt = NULL;
+   }
+   RETURN (interrupt);
+}
+
+void
+_mongoc_interrupt_destroy (mongoc_interrupt_t *interrupt)
+{
+   if (!interrupt) {
+      return;
+   }
+
+   mongoc_socket_destroy (interrupt->impl.socket_pair.read);
+   mongoc_socket_destroy (interrupt->impl.socket_pair.write);
+   mongoc_stream_destroy (interrupt->stream);
+   bson_mutex_destroy (&interrupt->mutex);
+   bson_free (interrupt);
+}
+
+bool
+_mongoc_interrupt_flush (mongoc_interrupt_t *interrupt)
+{
+   uint8_t buf[1];
+   while (true) {
+      if (-1 == mongoc_socket_recv (
+                   interrupt->impl.socket_pair.read, buf, sizeof (buf), 0, 0)) {
+         if (MONGOC_ERRNO_IS_AGAIN (errno)) {
+            /* Nothing left to read. */
+            break;
```

Comment:
Nit: For consistency, why not just return `true` here like the non-windows version?

Reply:
Ah true. Sara asked for the same change in the non-windows version: https://github.com/mongodb/mongo-c-driver/pull/610#discussion_r425292372

Done.